demos/sample_azure_iot_pnp/sample_azure_iot_pnp.c

```diff
@@ -0,0 +1,1035 @@
+static double xDeviceAverageTemperature = sampleazureiotDEFAULT_START_TEMP_CELSIUS;
+
+/**
+ * @brief The payload to send to the Device Provisioning Service
+ */
+#define sampleazureiotPROVISIONING_PAYLOAD                    "{\"modelId\":\"" sampleazureiotMODEL_ID "\"}"
+
+/**
+ * @brief Time in ticks to wait between each cycle of the demo implemented
+ * by prvMQTTDemoTask().
+ */
+#define sampleazureiotDELAY_BETWEEN_DEMO_ITERATIONS_TICKS     ( pdMS_TO_TICKS( 5000U ) )
+
+/**
+ * @brief Timeout for MQTT_ProcessLoop in milliseconds.
+ */
+#define sampleazureiotPROCESS_LOOP_TIMEOUT_MS                 ( 500U )
+
+/**
+ * @brief Delay (in ticks) between consecutive cycles of MQTT publish operations in a
+ * demo iteration.
+ *
+ * Note that the process loop also has a timeout, so the total time between
+ * publishes is the sum of the two delays.
+ */
+#define sampleazureiotDELAY_BETWEEN_PUBLISHES_TICKS           ( pdMS_TO_TICKS( 2000U ) )
+
+/**
+ * @brief Transport timeout in milliseconds for transport send and receive.
+ */
+#define sampleazureiotTRANSPORT_SEND_RECV_TIMEOUT_MS          ( 2000U )
+
+/**
+ * @brief Transport timeout in milliseconds for transport send and receive.
+ */
+#define sampleazureiotProvisioning_Registration_TIMEOUT_MS    ( 20U )
+
+/**
+ * @brief Wait timeout for subscribe to finish.
+ */
+#define sampleazureiotSUBSCRIBE_TIMEOUT                       ( 10 * 1000U )
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Unix time.
+ *
+ * @return Time in milliseconds.
+ */
+uint64_t ullGetUnixTime( void );
+/*-----------------------------------------------------------*/
+
+/* Define buffer for IoT Hub info.  */
+#ifdef democonfigENABLE_DPS_SAMPLE
+    static uint8_t ucSampleIotHubHostname[ 128 ];
+    static uint8_t ucSampleIotHubDeviceId[ 128 ];
+    static AzureIoTProvisioningClient_t xAzureIoTProvisioningClient;
+#endif /* democonfigENABLE_DPS_SAMPLE */
+
+static uint8_t ucScratchBuffer[ 128 ];
+
+/* Each compilation unit must define the NetworkContext struct. */
+struct NetworkContext
+{
+    TlsTransportParams_t * pParams;
+};
+
+static AzureIoTHubClient_t xAzureIoTHubClient;
+/*-----------------------------------------------------------*/
+
+#ifdef democonfigENABLE_DPS_SAMPLE
+
+/**
+ * @brief Gets the IoT Hub endpoint and deviceId from Provisioning service.
+ *   This function will block for Provisioning service for result or return failure.
+ *
+ * @param[in] pXNetworkCredentials  Network credential used to connect to Provisioning service
+ * @param[out] ppucIothubHostname  Pointer to uint8_t* IoT Hub hostname return from Provisioning Service
+ * @param[in,out] pulIothubHostnameLength  Length of hostname
+ * @param[out] ppucIothubDeviceId  Pointer to uint8_t* deviceId return from Provisioning Service
+ * @param[in,out] pulIothubDeviceIdLength  Length of deviceId
+ */
+    static uint32_t prvIoTHubInfoGet( NetworkCredentials_t * pXNetworkCredentials,
```

Comment:
>prvIoTHubInfoGet [](http://example.com/codeflow?start=20&length=16)

how about moving these common function between samples in common folder? and for all other generic functions?

Reply:
we can but tbh I'd rather not do it now. I can make a card and get to it?
https://github.com/Azure-Samples/iot-middleware-freertos-samples/issues/52